src/main/java/org/springframework/retry/support/RetryTemplate.java

```diff
@@ -529,12 +526,16 @@ private RetryContext doOpenInternal(RetryPolicy retryPolicy) {
 	 * @return T the payload to return
 	 * @throws Throwable if there is an error
 	 */
-	protected <T> T handleRetryExhausted(RecoveryCallback<T> recoveryCallback, RetryContext context, RetryState state)
-			throws Throwable {
+	protected <T> T handleRetryExhausted(RecoveryCallback<T> recoveryCallback, RetryContext context, RetryState state,
+			RetryPolicy retryPolicy) throws Throwable {
 		context.setAttribute(RetryContext.EXHAUSTED, true);
 		if (state != null && !context.hasAttribute(GLOBAL_STATE)) {
 			this.retryContextCache.remove(state.getKey());
 		}
+		if (this.throwLastExceptionOnExhausted && retryPolicy instanceof SimpleRetryPolicy
```

Comment:
This is a behavior change and could break existing projects that expect the recovery callback to be invoked for all exceptions. I think we need a new boolean `noRecoveryForNotRetryable` or similar.

Reply:
`throwLastExceptionOnExhausted` is true only if `Retryable.rethrow` is true, which is false per default. So, all the old code will have `throwLastExceptionOnExhausted` set to false and this new `if` branch will be skipped. 

Could you please collaborate more if I miss something? Thanks!

Reply:
@ttulka Sorry for the delay. The problem is for users of the `RetryTemplate` (those not using `@Retryable`, which is very common).

If an existing application is using the template with `throwLastExceptionOnExhausted` then the recovery callback is still invoked.

With your change, it will no longer be called; hence the new behavior needs to be controlled by a new property on the template rather than overloading the existing property to have 2 meanings.

Reply:
@garyrussell Thanks for clearing things up! I have committed the changes with the new flag.